### DISCUSSION TITLE
Explicit Call to spdlog::shutdown Required?

### DISCUSSION THREAD

**Discussion by mattbrown015** [gabime/spdlog#2685] [Q&A] [answered]
Hi,

I'm not currently explicitly calling `spdlog::shudown` because I wasn't sure where to call it from.

Our application is large, with a horrendous mass of threads!  We've always had problems closing and have long suspected we've got deadlocks, or some such, when trying to terminate all the threads.  Recently I've got the feeling that it has got worse and that feeling coincides with my attempts to introduce spdlog.

I've created spdlogger instances with async sinks and configured timed log flush, which I believe means I'm asking spdlog to create yet more threads.

I looked at our main and considered adding `spdlog::shutdown` to the end but didn't because I assume some threads may still be running, and possibly logging, at this point.

Should I add `spdlog::shutdown` to the end of main in a multithreaded application?

We use spdlog as a shared library.  The application runs on Windows and Linux.

Clearly I could, and should, investigated what's preventing our app closing correctly but it isn't consistent and doesn't cause our customers much pain hence there are always more important things to do.

Thanks,
Matt

**Comment by tt4g** [comment, marked as answer]
> Should I add spdlog::shutdown to the end of main in a multithreaded application?
> 
> We use spdlog as a shared library. The application runs on Windows and Linux.

Yes.
spdlog creates a thread pool for the async logger.
It is recommended to call `spdlog::shutdown()` as it destroys the thread pool.

And `spdlog::shutdown()` calls sink's flush operation. This operation is important on Windows because Windows discards the data remaining in the I/O buffer after `main()`.

However, I will tell you that it is important to properly clean up all threads of your application.
This is because `std::thread` calls `std::thread::terminate()` in the destructor unless `std::thread::join()` or `std::thread::detach()` is called.
See; [c\+\+ \- C\+\+11: What happens if you don't call join\(\) for std::thread \- Stack Overflow](https://stackoverflow.com/questions/27392743/c11-what-happens-if-you-dont-call-join-for-stdthread)

Even if you don't call `spdlog::shutdown()`, the OS may call the destructor of the spdlog created resource after `main()`, but if `std::terminate()` is called in the `std::thread` destructor, everything is ruined.

**Reply by mattbrown015** [reply]
> It is recommended to call spdlog::shutdown() as it destroys the thread pool.

Yes, I ok, thanks for your help.  It's good to talk and I think my original reasoning was floored!  What you're saying is that we should have joined all the threads before we get to the end of main.  Of course we should!  And we do try!  So I will add the shutdown.

> However, I will tell you that it is important to properly clean up all threads of your application.

Yes, we do appreciate that and some of us have been trying to improve things but it's an old code base and clearly we've got problems that are hard to track down! :-(

Part of the reason for converting to spdlog is that there's an old inhouse logging solution that isn't thread safe!

spdlog is fantastic, BTW! Thanks to the community! :-)

**Comment by BunnySakura** [comment]
U can try `std::at_exit`, and call `shutdown` in this.
